src/v/archival/ntp_archiver_service.cc

```diff
@@ -71,6 +71,33 @@ constexpr auto housekeeping_jit = 5ms;
 
 namespace archival {
 
+static bool segment_meta_matches_stats(
+  const cloud_storage::segment_meta& meta,
+  const cloud_storage::segment_record_stats& stats,
+  retry_chain_logger& ctxlog) {
+    // Validate segment content. The 'stats' is computed when
+    // the actual segment is scanned and represents the 'ground truth' about
+    // its content. The 'meta' is the expected segment metadata. We
+    // shouldn't replicate it if it doesn't match the 'stats'.
+    if (
```

Comment:
We could also check the deltas to some degree:
`meta.delta_offest_end - meta.delta_offset  != stats.total_conf_records`

Reply:
done